File: app/src/main/java/com/example/changfeng/taptapword/FileListActivity.java
package com.example.changfeng.taptapword;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.os.Environment;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Toast;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by changfeng on 2015/3/4.
 */
public class FileListActivity extends Activity implements AdapterView.OnItemClickListener {

    private static final String TAG = "FileListActivity";
    public static final String FILE_KEY_WORD = "word_ninja";

    List<FileInfo> fileInfoList;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        requestWindowFeature(Window.FEATURE_NO_TITLE);

//        Log.d(TAG, "onCreate() called");
        setContentView(R.layout.activity_file_list);


        Intent intent = getIntent();
        fileInfoList = getFileList(intent.getStringExtra(FILE_KEY_WORD));
        if (fileInfoList.isEmpty()) {
            fileInfoList.add(new FileInfo(getString(R.string.filename_no_backup_file_found), getString(R.string.filepath_no_backup_file_found)));
        }

        final FileListAdapter adapter = new FileListAdapter(FileListActivity.this,
                R.layout.file_list_item, fileInfoList);

        ListView listView = (ListView) findViewById(R.id.file_list_view);
        listView.setAdapter(adapter);
        listView.setOnItemClickListener(this);

    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        FileInfo fileInfo = fileInfoList.get(position);
        if (!fileInfo.fileName.equals(getString(R.string.filename_no_backup_file_found))) {
            setResults(fileInfo);
        }
        finish();
    }

    private void setResults(FileInfo fileInfo) {
        Intent intent = new Intent();
        intent.putExtra(FILE_KEY_WORD, fileInfo.filePath);
        setResult(Activity.RESULT_OK, intent);
//        Log.d(TAG, "onItemClick() called filepath:" + fileInfo.filePath);
    }

    private ArrayList<FileInfo> getFileList(String fileKeyWord) {
        ArrayList<FileInfo> list = new ArrayList<>();
        if (Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED)) {
            File path = Environment.getExternalStorageDirectory();
            File files[] = path.listFiles();
            if (files != null) {
                for (File file : files) {
                    if (file.isDirectory()) {
                        continue;
                    }
                    if (file.getName().contains(fileKeyWord)) {
                        list.add(new FileInfo(file.getName(), file.getPath()));
                    }
                }
            }
        }
        return list;
    }
}

File: app/src/main/java/com/example/changfeng/taptapword/FileListActivity.kt
package com.example.changfeng.taptapword

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.os.Environment
import android.view.Window
import android.widget.ListView
import com.umeng.analytics.MobclickAgent
import org.jetbrains.anko.find
import org.jetbrains.anko.sdk27.coroutines.onItemClick

/**
 * Created by changfeng on 2015/3/4.
 */
class FileListActivity : Activity() {

    var fileInfoList: MutableList<FileInfo>? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        requestWindowFeature(Window.FEATURE_NO_TITLE)
        setContentView(R.layout.activity_file_list)

        fileInfoList = getFileList(intent.getStringExtra(FILE_KEY_WORD))
        fileInfoList?.let {
            if (fileInfoList!!.isEmpty()) {
                fileInfoList!!.add(FileInfo(getString(R.string.filename_no_backup_file_found), getString(R.string.filepath_no_backup_file_found)))
            }
        }


        val adapter = FileListAdapter(this@FileListActivity,
            R.layout.file_list_item, fileInfoList!!)

        val listView = find<ListView>(R.id.file_list_view)
        listView.adapter = adapter
        listView.onItemClick { _, _, i, _ ->
            fileInfoList?.let {
                val fileInfo = fileInfoList!![i]
                if (fileInfo.fileName != getString(R.string.filename_no_backup_file_found)) {
                    setResults(fileInfo)
                }
            }
            finish()
        }

    }


    override fun onResume() {
        super.onResume()
        MobclickAgent.onPageStart("选择文件页面")
        MobclickAgent.onResume(this)
    }

    override fun onPause() {
        super.onPause()
        MobclickAgent.onPageEnd("选择文件页面")
        MobclickAgent.onPause(this)
    }

    private fun setResults(fileInfo: FileInfo) {
        setResult(Activity.RESULT_OK, Intent().putExtra(FILE_KEY_WORD, fileInfo.filePath))
    }

    private fun getFileList(fileKeyWord: String): MutableList<FileInfo> {
        if (Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED) {
            return Environment.getExternalStorageDirectory()
                .listFiles { _, s -> s.contains(fileKeyWord) }
                .filter { s -> s.isFile }
                .map { FileInfo(it.name, it.path) }.toMutableList()
        } else {
            return mutableListOf()
        }
    }

    companion object {

        private val TAG = "FileListActivity"
        val FILE_KEY_WORD = "word_ninja"
    }
}
